File: java/designpattern/behavioral/Observer.java
package designpattern.behavioral;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Observer Pattern Example
 * 
 * Intent: Define a one-to-many dependency between objects so that when one object changes state,
 * all its dependents are notified and updated automatically.
 * 
 * Real-world analogy:
 * Think of a payment processing system where multiple components need to be notified
 * about payment status changes:
 * 1. Customer notifications (email, SMS)
 * 2. Merchant notifications
 * 3. Payment gateway updates
 * 4. Analytics tracking
 * 5. Fraud detection system
 */

// Subject (Observable)
abstract class PaymentSubject {
    protected Set<PaymentObserver> observers;

    public PaymentSubject() {
        this.observers = new HashSet<>();
    }

    public void attach(PaymentObserver observer) {
        observers.add(observer);
        System.out.println("Observer attached: " + observer.getClass().getSimpleName());
    }

    public void detach(PaymentObserver observer) {
        observers.remove(observer);
        System.out.println("Observer detached: " + observer.getClass().getSimpleName());
    }

    protected void notify(Payment payment, String event) {
        for (PaymentObserver observer : observers) {
            observer.update(payment, event);
        }
    }
}

// Concrete Subject
class PaymentProcessor extends PaymentSubject {
    private Map<String, Payment> payments;

    public PaymentProcessor() {
        super();
        this.payments = new ConcurrentHashMap<>();
    }

    public void processPayment(Payment payment) {
        System.out.println("\nProcessing payment...");
        payments.put(payment.getTransactionId(), payment);
        
        // Notify observers about payment initiation
        notify(payment, "PAYMENT_INITIATED");
        
        if (payment.validate()) {
            payment.process();
            notify(payment, "PAYMENT_PROCESSED");
        } else {
            payment.setStatus(PaymentStatus.FAILED);
            notify(payment, "PAYMENT_FAILED");
        }
    }

    public void refundPayment(Payment payment) {
        System.out.println("\nProcessing refund...");
        if (payment.getStatus() == PaymentStatus.COMPLETED) {
            payment.refund();
            notify(payment, "PAYMENT_REFUNDED");
        } else {
            notify(payment, "REFUND_FAILED");
        }
    }
}

// Observer interface
interface PaymentObserver {
    void update(Payment payment, String event);
}

// Concrete Observers
class CustomerNotificationObserver implements PaymentObserver {
    private Set<String> notifications;

    public CustomerNotificationObserver() {
        this.notifications = new HashSet<>();
    }

    @Override
    public void update(Payment payment, String event) {
        String message = createNotificationMessage(payment, event);
        notifications.add(message);
        System.out.println("\nCustomer Notification: " + message);
        
        // In a real system, this would send actual notifications
        switch (event) {
            case "PAYMENT_PROCESSED":
                sendEmail(payment, "Payment Successful");
                sendSMS(payment, "Payment Successful");
                break;
            case "PAYMENT_FAILED":
                sendEmail(payment, "Payment Failed");
                sendSMS(payment, "Payment Failed");
                break;
            case "PAYMENT_REFUNDED":
                sendEmail(payment, "Payment Refunded");
                sendSMS(payment, "Payment Refunded");
                break;
        }
    }

    private String createNotificationMessage(Payment payment, String event) {
        return String.format("%s: %.2f %s - %s", 
            event, payment.getAmount(), payment.getCurrency(), payment.getStatus());
    }

    private void sendEmail(Payment payment, String subject) {
        System.out.println("Sending email to customer: " + subject);
    }

    private void sendSMS(Payment payment, String message) {
        System.out.println("Sending SMS to customer: " + message);
    }
}

class MerchantNotificationObserver implements PaymentObserver {
    @Override
    public void update(Payment payment, String event) {
        System.out.println("\nMerchant Notification: " + event + " for payment " + payment.getTransactionId());
        System.out.println("Amount: " + payment.getAmount() + " " + payment.getCurrency());
        System.out.println("Status: " + payment.getStatus());
        
        // In a real system, this would update merchant dashboard
        if (event.equals("PAYMENT_PROCESSED")) {
            updateMerchantDashboard(payment);
        }
    }

    private void updateMerchantDashboard(Payment payment) {
        System.out.println("Updating merchant dashboard with new payment");
    }
}

class FraudDetectionObserver implements PaymentObserver {
    private Set<String> suspiciousTransactions;

    public FraudDetectionObserver() {
        this.suspiciousTransactions = new HashSet<>();
    }

    @Override
    public void update(Payment payment, String event) {
        if (isSuspiciousTransaction(payment)) {
            suspiciousTransactions.add(payment.getTransactionId());
            System.out.println("\nFraud Alert: Suspicious transaction detected!");
            System.out.println("Transaction ID: " + payment.getTransactionId());
            System.out.println("Amount: " + payment.getAmount() + " " + payment.getCurrency());
            notifyFraudTeam(payment);
        }
    }

    private boolean isSuspiciousTransaction(Payment payment) {
        // Simple fraud detection logic
        return payment.getAmount() > 10000 || 
               (payment.getPaymentMethod() == PaymentMethod.CREDIT_CARD && payment.getAmount() > 5000);
    }

    private void notifyFraudTeam(Payment payment) {
        System.out.println("Notifying fraud detection team...");
    }
}

class AnalyticsObserver implements PaymentObserver {
    private static class Analytics {
        int totalProcessed;
        double totalAmount;
        int failedPayments;
        int refundedPayments;
    }

    private Analytics analytics;

    public AnalyticsObserver() {
        this.analytics = new Analytics();
    }

    @Override
    public void update(Payment payment, String event) {
        switch (event) {
            case "PAYMENT_PROCESSED":
                analytics.totalProcessed++;
                analytics.totalAmount += payment.getAmount();
                break;
            case "PAYMENT_FAILED":
                analytics.failedPayments++;
                break;
            case "PAYMENT_REFUNDED":
                analytics.refundedPayments++;
                analytics.totalAmount -= payment.getAmount();
                break;
        }
        
        displayAnalytics();
    }

    private void displayAnalytics() {
        System.out.println("\nPayment Analytics:");
        System.out.println("-----------------");
        System.out.println("Total Processed: " + analytics.totalProcessed);
        System.out.println("Total Amount: " + analytics.totalAmount);
        System.out.println("Failed Payments: " + analytics.failedPayments);
        System.out.println("Refunded Payments: " + analytics.refundedPayments);
    }
}

// Test the pattern
public class Observer {
    public static void main(String[] args) {
        System.out.println("Observer Pattern Demo - Payment Notifications\n");
        
        // Create subject
        PaymentProcessor paymentProcessor = new PaymentProcessor();
        
        // Create observers
        PaymentObserver customerNotifier = new CustomerNotificationObserver();
        PaymentObserver merchantNotifier = new MerchantNotificationObserver();
        PaymentObserver fraudDetector = new FraudDetectionObserver();
        PaymentObserver analyticsTracker = new AnalyticsObserver();
        
        // Attach observers
        paymentProcessor.attach(customerNotifier);
        paymentProcessor.attach(merchantNotifier);
        paymentProcessor.attach(fraudDetector);
        paymentProcessor.attach(analyticsTracker);
        
        // Create and process payments
        Payment payment1 = new CreditCardPayment(500.0, "USD", "1234567890123456", "12/25", "123");
        Payment payment2 = new CreditCardPayment(15000.0, "USD", "9876543210987654", "12/25", "456");
        
        // Process payments
        paymentProcessor.processPayment(payment1);
        paymentProcessor.processPayment(payment2);
        
        // Test refund
        paymentProcessor.refundPayment(payment1);
        
        // Detach an observer
        paymentProcessor.detach(merchantNotifier);
        
        // Process another payment (merchant won't be notified)
        Payment payment3 = new PayPalPayment(1000.0, "USD", "test@example.com");
        paymentProcessor.processPayment(payment3);
    }
} 

File: javascript/designpattern/behavioral/Observer.js
/**
 * Observer Pattern Example
 * 
 * Intent: Define a one-to-many dependency between objects so that when one object changes state,
 * all its dependents are notified and updated automatically.
 * 
 * Real-world analogy:
 * Think of a payment processing system where multiple components need to be notified
 * about payment status changes:
 * 1. Customer notifications (email, SMS)
 * 2. Merchant notifications
 * 3. Payment gateway updates
 * 4. Analytics tracking
 * 5. Fraud detection system
 */

const { Payment, PaymentStatus } = require('./PaymentTheme');

// Subject (Observable)
class PaymentSubject {
    constructor() {
        this.observers = new Set();
    }

    attach(observer) {
        this.observers.add(observer);
        console.log(`Observer attached: ${observer.constructor.name}`);
    }

    detach(observer) {
        this.observers.delete(observer);
        console.log(`Observer detached: ${observer.constructor.name}`);
    }

    notify(payment, event) {
        for (const observer of this.observers) {
            observer.update(payment, event);
        }
    }
}

// Concrete Subject
class PaymentProcessor extends PaymentSubject {
    constructor() {
        super();
        this.payments = new Map();
    }

    processPayment(payment) {
        console.log('\nProcessing payment...');
        this.payments.set(payment.transactionId, payment);
        
        // Notify observers about payment initiation
        this.notify(payment, 'PAYMENT_INITIATED');
        
        if (payment.validate()) {
            payment.process();
            this.notify(payment, 'PAYMENT_PROCESSED');
        } else {
            payment.status = PaymentStatus.FAILED;
            this.notify(payment, 'PAYMENT_FAILED');
        }
    }

    refundPayment(payment) {
        console.log('\nProcessing refund...');
        if (payment.status === PaymentStatus.COMPLETED) {
            payment.refund();
            this.notify(payment, 'PAYMENT_REFUNDED');
        } else {
            this.notify(payment, 'REFUND_FAILED');
        }
    }
}

// Observer interface
class PaymentObserver {
    update(payment, event) {
        throw new Error('Method update() must be implemented');
    }
}

// Concrete Observers
class CustomerNotificationObserver extends PaymentObserver {
    constructor() {
        super();
        this.notifications = [];
    }

    update(payment, event) {
        const message = this.createNotificationMessage(payment, event);
        this.notifications.push(message);
        console.log(`\nCustomer Notification: ${message}`);
        
        // In a real system, this would send actual notifications
        switch (event) {
            case 'PAYMENT_PROCESSED':
                this.sendEmail(payment, 'Payment Successful');
                this.sendSMS(payment, 'Payment Successful');
                break;
            case 'PAYMENT_FAILED':
                this.sendEmail(payment, 'Payment Failed');
                this.sendSMS(payment, 'Payment Failed');
                break;
            case 'PAYMENT_REFUNDED':
                this.sendEmail(payment, 'Payment Refunded');
                this.sendSMS(payment, 'Payment Refunded');
                break;
        }
    }

    createNotificationMessage(payment, event) {
        return `${event}: ${payment.amount} ${payment.currency} - ${payment.status}`;
    }

    sendEmail(payment, subject) {
        console.log(`Sending email to customer: ${subject}`);
    }

    sendSMS(payment, message) {
        console.log(`Sending SMS to customer: ${message}`);
    }
}

class MerchantNotificationObserver extends PaymentObserver {
    update(payment, event) {
        console.log(`\nMerchant Notification: ${event} for payment ${payment.transactionId}`);
        console.log(`Amount: ${payment.amount} ${payment.currency}`);
        console.log(`Status: ${payment.status}`);
        
        // In a real system, this would update merchant dashboard
        if (event === 'PAYMENT_PROCESSED') {
            this.updateMerchantDashboard(payment);
        }
    }

    updateMerchantDashboard(payment) {
        console.log('Updating merchant dashboard with new payment');
    }
}

class FraudDetectionObserver extends PaymentObserver {
    constructor() {
        super();
        this.suspiciousTransactions = new Set();
    }

    update(payment, event) {
        if (this.isSuspiciousTransaction(payment)) {
            this.suspiciousTransactions.add(payment.transactionId);
            console.log(`\nFraud Alert: Suspicious transaction detected!`);
            console.log(`Transaction ID: ${payment.transactionId}`);
            console.log(`Amount: ${payment.amount} ${payment.currency}`);
            this.notifyFraudTeam(payment);
        }
    }

    isSuspiciousTransaction(payment) {
        // Simple fraud detection logic
        return payment.amount > 10000 || 
               (payment.paymentMethod === 'CREDIT_CARD' && payment.amount > 5000);
    }

    notifyFraudTeam(payment) {
        console.log('Notifying fraud detection team...');
    }
}

class AnalyticsObserver extends PaymentObserver {
    constructor() {
        super();
        this.analytics = {
            totalProcessed: 0,
            totalAmount: 0,
            failedPayments: 0,
            refundedPayments: 0
        };
    }

    update(payment, event) {
        switch (event) {
            case 'PAYMENT_PROCESSED':
                this.analytics.totalProcessed++;
                this.analytics.totalAmount += payment.amount;
                break;
            case 'PAYMENT_FAILED':
                this.analytics.failedPayments++;
                break;
            case 'PAYMENT_REFUNDED':
                this.analytics.refundedPayments++;
                this.analytics.totalAmount -= payment.amount;
                break;
        }
        
        this.displayAnalytics();
    }

    displayAnalytics() {
        console.log('\nPayment Analytics:');
        console.log('-----------------');
        console.log(`Total Processed: ${this.analytics.totalProcessed}`);
        console.log(`Total Amount: ${this.analytics.totalAmount}`);
        console.log(`Failed Payments: ${this.analytics.failedPayments}`);
        console.log(`Refunded Payments: ${this.analytics.refundedPayments}`);
    }
}

// Test the pattern
function testObserver() {
    console.log('Observer Pattern Demo - Payment Notifications\n');
    
    // Create subject
    const paymentProcessor = new PaymentProcessor();
    
    // Create observers
    const customerNotifier = new CustomerNotificationObserver();
    const merchantNotifier = new MerchantNotificationObserver();
    const fraudDetector = new FraudDetectionObserver();
    const analyticsTracker = new AnalyticsObserver();
    
    // Attach observers
    paymentProcessor.attach(customerNotifier);
    paymentProcessor.attach(merchantNotifier);
    paymentProcessor.attach(fraudDetector);
    paymentProcessor.attach(analyticsTracker);
    
    // Create and process payments
    const payment1 = new Payment(500, 'USD', 'CREDIT_CARD');
    payment1.cardNumber = '1234567890123456';
    payment1.expiryDate = '12/25';
    payment1.cvv = '123';
    
    const payment2 = new Payment(15000, 'USD', 'CREDIT_CARD');
    payment2.cardNumber = '9876543210987654';
    payment2.expiryDate = '12/25';
    payment2.cvv = '456';
    
    // Process payments
    paymentProcessor.processPayment(payment1);
    paymentProcessor.processPayment(payment2);
    
    // Test refund
    paymentProcessor.refundPayment(payment1);
    
    // Detach an observer
    paymentProcessor.detach(merchantNotifier);
    
    // Process another payment (merchant won't be notified)
    const payment3 = new Payment(1000, 'USD', 'PAYPAL');
    payment3.email = 'test@example.com';
    paymentProcessor.processPayment(payment3);
}

// Run the test
testObserver(); 